### PR TITLE
fix: prevent crash on out-of-range integer literals (#729)

### LINKED ISSUE #729: fix: parser crashes on int64 max literal (9223372036854775807) with stoll out_of_range
## Bug

The parser crashes with `libc++abi: terminating due to uncaught exception of type std::out_of_range: stoll: out of range` when the int64 max value is used as a literal.

## Reproducer

```ry
max_val = 9223372036854775807   # i64 max
```

Causes abort with uncaught `std::out_of_range` from `stoll`.

## Root Cause

`ry::parseIntLiteral` calls `std::stoll` without catching `std::out_of_range`. When the literal value equals `INT64_MAX` (or `INT64_MIN`), `stoll` throws an exception that terminates the process.

Note: `INT64_MAX + 1 = 9223372036854775808` would overflow, but `9223372036854775807` is the exact max and should be valid.

## Expected Behavior

Should either:
1. Parse successfully and store as `i64` value
2. Give a clear compile-time error ("integer literal out of range for int")

## Context

Discovered during combinatorial stdlib boundary test implementation (#628).

### PR DESCRIPTION
**Pull request by t0k0sh1**
$(cat <<'EOF'
## Summary

- `std::stoll` throws `std::out_of_range` for integers outside the `int64` range, which propagated as an uncaught exception and crashed the parser
- Replaced with `strtoll` + `errno` for safe range detection
- Out-of-range literals now produce a clean compile error instead of a crash

## Test plan

- [ ] `IntLiteralInt64Max`: `9223372036854775807` parses to `INT64_MAX`
- [ ] `IntLiteralOutOfRangeThrows`: `9223372036854775808` produces `DiagnosticError` with "integer literal out of range"
- [ ] `./build/ry_tests` passes
- [ ] `./build/ry test -p` passes
- [ ] ASan build passes

Closes #729
EOF
)

## Summary by CodeRabbit

* **Bug Fixes**
  * Integer literal parser now handles out-of-range values gracefully, emitting a clear compile error instead of crashing.

* **Tests**
  * Added tests validating integer literal boundary behavior and error handling for edge cases.

### CODE REVIEW

File: include/ry/parser.hpp [hunk 1]
```diff
@@ -151,13 +151,28 @@ inline std::string stripUnderscores(const std::string &s) {
...
+    const char *p = clean.c_str();
     if (clean.size() > 2 && clean[0] == '0') {
-        if (clean[1] == 'x' || clean[1] == 'X') return std::stoll(clean, nullptr, 16);
-        if (clean[1] == 'b' || clean[1] == 'B') return std::stoll(clean.substr(2), nullptr, 2);
+        if (clean[1] == 'x' || clean[1] == 'X') base = 16;
+        else if (clean[1] == 'b' || clean[1] == 'B') { base = 2; p += 2; }
     }
-    return std::stoll(clean);
+    long long val = std::strtoll(p, &end, base);
+    if (errno == ERANGE || (end && *end != '\0')) return false;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply include/ry/parser.hpp:169]: Fixed: added `<cerrno>`, `<cstdlib>`, and `<stdexcept>` to `parser.hpp` includes.

File: tests/spec/combinatorial/syntax_combinations.test.ry [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    expect(first_of_one((99,))).to_eq(99)
+  )
+)
+
+describe("syntax combo — chained index/method access", ():
+  it("should chain map index and list index access", ():
+    m: Map<str, List<int>> = {"nums": [10, 20, 30]}
+    xs = m["nums"]
+    expect(xs[1]).to_eq(20)
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/spec/combinatorial/syntax_combinations.test.ry]: Fixed: changed to `expect(m["nums"][1]).to_eq(20)` so the chained index access is inline and actually exercises the codegen path.

File: src/parser_expr.cpp [hunk 3]
```diff
@@ -288,8 +288,13 @@ ExprPtr Parser::parsePrimary() {
...
         if (suffix == "f32" || suffix == "f64")
             node->data = FloatExpr{parseFloatLiteral(numStr), suffix};
-        else
-            node->data = NumberExpr{parseIntLiteral(numStr), suffix};
+        else {
+            try {
+                node->data = NumberExpr{parseIntLiteral(numStr), suffix};
+            } catch (const std::out_of_range &e) {
+                parseError(t.line, e.what());
+            }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/parser_expr.cpp]: Fixed: restructured to validate with `tryParseIntLiteral` before calling `lex_.next()`, so `parseError()` is called while the number token is still the peek position, giving an accurate column in diagnostics.

File: src/parser_decl.cpp [hunk 4]
```diff
@@ -889,8 +899,13 @@ Pattern Parser::parsePattern() {
...
             if (suffix == "f32" || suffix == "f64")
                 node->data = FloatExpr{-parseFloatLiteral(numStr), suffix};
-            else
-                node->data = NumberExpr{-parseIntLiteral(numStr), suffix};
+            else {
+                try {
+                    node->data = NumberExpr{-parseIntLiteral(numStr), suffix};
+                } catch (const std::out_of_range &e) {
+                    parseError(num.line, e.what());
+                }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/parser_decl.cpp]: Fixed: same restructuring for the negative literal pattern path.

File: src/parser_decl.cpp [hunk 5]
```diff
@@ -575,7 +575,12 @@ StmtNode Parser::parseEnumStatement() {
             if (valTok.kind != TokenKind::Number)
                 parseError(valTok.line, "expected integer literal for enum variant value");
             lex_.next(); // consume number
-            int64_t val = std::stoll(valTok.value);
+            int64_t val;
+            try {
+                val = std::stoll(valTok.value);
+            } catch (const std::out_of_range &) {
+                parseError(valTok.line, "integer literal out of range for int: " + valTok.value);
+            }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/parser_decl.cpp]: Fixed: replaced `std::stoll` with `tryParseIntLiteral` so enum explicit values follow the same literal parsing rules (0x/0b/0o prefixes, underscore separators) as other integer literals.

File: tests/spec/combinatorial/syntax_combinations.test.ry [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    expect(label).to_eq("three")
+  )
+  it("should use when expression in function argument", ():
+    n = 5
+    result = (when:
+      n > 10 => n * 2
+      n > 3 => n + 10
+      else => n
+    )
+    expect(result).to_eq(15)
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/spec/combinatorial/syntax_combinations.test.ry]: Passing `when:` directly as a function argument is not yet supported by the current parser (only parenthesized `when` or `when` in statement position works). Renamed the test to "should use when expression in expression position" and added a comment documenting the limitation.

File: src/parser_decl.cpp [hunk 7]
```diff
@@ -855,8 +860,13 @@ Pattern Parser::parsePattern() {
...
         if (suffix == "f32" || suffix == "f64")
             node->data = FloatExpr{parseFloatLiteral(numStr), suffix};
-        else
-            node->data = NumberExpr{parseIntLiteral(numStr), suffix};
+        else {
+            try {
+                node->data = NumberExpr{parseIntLiteral(numStr), suffix};
+            } catch (const std::out_of_range &e) {
+                parseError(t.line, e.what());
+            }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/parser_decl.cpp]: Fixed: same restructuring as parser_expr.cpp — validate before consuming the token so error location points at the literal.

File: tests/spec/combinatorial/print_display.test.ry [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# BUG #728: print() and to_str() on closure values not supported
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/spec/combinatorial/print_display.test.ry]: Already fixed by the v0.0.8 merge: the comment now reads "BUG #728: print() and to_str() on int | str union values not supported".

File: changelog.d/628-combinatorial-coverage.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+### Added
+
+- Systematic combinatorial test coverage in `tests/spec/combinatorial/` (#628): 113 tests across 9 files covering type×operation matrix (equality, fn argument/return, collection element, match, nested types, syntax combinations, print/display, stdlib boundary inputs)
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply changelog.d/628-combinatorial-coverage.md:3]: Fixed: added `changelog.d/729-int64-max.md` describing the out-of-range integer literal fix. The `628-combinatorial-coverage.md` file was already merged into v0.0.8 and is a no-op in this PR.
